### PR TITLE
Blog: Advanced Git with the Free University of Tbilisi

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
My speaking notes from joining @iomodo today and his students at the Free University of Tbilisi.

See https://docs.google.com/presentation/d/1tVH2kw_WJ5_rCCOIGHPxVc8O07g1zHK0hQIloCwEYAk/edit#slide=id.g8175639bd3_0_52 for the original slides, and https://github.com/mattermost/mattermost-server/pull/14094 for the associated mattermost-server changes (might as well actually commit them!)

### CODE REVIEW

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+github: lieut-data
+community: jesse.hallam
+---
+
+This morning, I had the privilege to (virtually) join [Shota Gvinepadze](https://community.mattermost.com/core/messages/@shota.gvinepadze) and his students at the [Free University of Tbilisi](http://freeuni.edu.ge/en) and speak about "Advanced Git @ Mattermost" for a portion of their class time. The [slides](https://docs.google.com/presentation/d/1tVH2kw_WJ5_rCCOIGHPxVc8O07g1zHK0hQIloCwEYAk/edit) are available now, and a recording may be made available soon.
+
+The following are my speaking notes from the session, slightly modified for this format. Keep in mind that the command line examples are illustrative of my workflow, and not meant to be run in isolation.
+
+### Introduction
+Today's session on open-source will focus on understanding Git better. I've interacted with a lot of people who are "scared of Git." They know the basics -- pulling, committing, pushing -- but anytime something goes wrong, they're stuck. Or they know about some of the advanced Git commands, but worry every time they have to do one.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
Today's session on open source will focus on understanding Git better. I've interacted with a lot of people who are "scared of Git." They know the basics -- pulling, committing, pushing -- but anytime something goes wrong, they're stuck. Or they know about some of the advanced Git commands, but worry every time they have to do one.
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+
+Over the next hour, my hope is to demystify some of these advanced commands, and show you how I use these tools to solve day-to-day problems while working at Mattermost.
+
+### Advanced Git
+
+What do I mean by advanced git? I'm thinking of four different operations in particular:
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
What do I mean by advanced Git? I'm thinking of four different operations in particular:
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+
+Just over a month later, one of larger customers upgraded to v4.9 and found their system grinding to a halt. We hopped on an emergency call to glean some information, and everything pointed to my changes making things much worse instead of much better. You can read the [details of my second investigation](https://community.mattermost.com/core/pl/n8bfq7wq77rkff8mgrrstqeuye) into what went wrong, but in the end, we decided to [revert my changes](https://github.com/mattermost/mattermost-server/pull/8659) and ship a patch release.
+
+When I undid these changes, I wanted to bring the code back to exactly the way it was before my changes. We already knew how that code behaved for this customer, and the goal was to restore stability by using the old code instead. I could, of course, make the changes manually: finding the old code and copying it over top of the new code. But one of git's built in commands is an operation to do this automatically. Let's see it in action!
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
When I undid these changes, I wanted to bring the code back to exactly the way it was before my changes. We already knew how that code behaved for this customer, and the goal was to restore stability by using the old code instead. I could, of course, make the changes manually: finding the old code and copying it over top of the new code. But one of Git's built-in commands is an operation to do this automatically. Let's see it in action!
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+* You can revert a revert: it's just like any other commit.
+* You can push your branch, containing the revert, and submit a pull request to review your changes.
+* If you're working out of a shared branch, your peers will pull the revert like any other commit. This is usually the right way to "undo changes" in a shared branch, vs. using a command that rewrites history and requires a force push instead.
+
+Of course, if you've ever done this before, you know that I've kind of cheated today: my revert was clean, in that there were no merge conflicts. Sometimes, especially if some time has passed, other changes might have touched the same lines, and the revert will require resolving those changes. Hold that thought, though, since I'll work through an example of resolving merge conflicts with the next command. 
+Remember, reverts are just new commits!
+
+### git cherry-pick
+
+Next up: cherry picking! The general idea of cherry picking -- not specific to git -- is to pick and choose just what you want. You can imagine yourself standing in front of an actual cherry tree, and saying, "I'll take that one, and that one, and that one," and ignoring the rest.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
Next up: cherry picking! The general idea of cherry picking -- not specific to Git -- is to pick and choose just what you want. You can imagine yourself standing in front of an actual cherry tree, and saying, "I'll take that one, and that one, and that one" and ignoring the rest.
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+
+In git, the idea is similar. Given one or more commits -- anywhere in the tree -- apply those same changes to my current ref. Why would you want to do that?
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
In Git, the idea is similar. Given one or more commits -- anywhere in the tree -- apply those same changes to my current ref. Why would you want to do that?
```

Comment (justinegeffen) [reply site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]: Question: Is "ref" meant to be "repo"?

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+
+At Mattermost, master is our development branch. Every two months we [release a new version](https://developers.mattermost.com/contribute/getting-started/branching/) of Mattermost that branches off master. Our release team begins to qualify this release, inevitably finding issues. We first fix the bugs in master, and then cherry-pick those changes back to the release branch. Of course, there are lots of different ways of using Git to handle releases, but this is the strategy we use at Mattermost.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
At Mattermost, `master` is our development branch. Every two months we [release a new version](https://developers.mattermost.com/contribute/getting-started/branching/) of Mattermost that branches off `master`. Our release team begins to qualify this release, inevitably finding issues. We first fix the bugs in `master`, and then cherry-pick those changes back to the `release` branch. Of course, there are lots of different ways of using Git to handle releases, but this is the strategy we use at Mattermost.
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+```sh
+# Checkout the mattermost-server repository (if not already done)
+git checkout https://github.com/mattermost/mattermost-server.git
+cd mattermost-server
+
+# Examine the combined history of release-5.20 and master
+git log --graph --oneline release-5.20 master
+```
+
+Observe that master has advanced in the weeks since we released v5.20. When we first released v5.20.0, a community member tried installing it and promptly [reported](https://community.mattermost.com/core/pl/877mkcc967ytjpgcgpbo6bi7ty) that his server was crashing! A bit of debugging on a Saturday, and I realized that we had introduced a regression into a code path we didn't often use ourselves. I filed [MM-21619](https://mattermost.atlassian.net/browse/MM-22619) and got to work on the fix. Let's find it in the master branch.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
Observe that `master` has advanced in the weeks since we released v5.20. When we first released v5.20.0, a community member tried installing it and promptly [reported](https://community.mattermost.com/core/pl/877mkcc967ytjpgcgpbo6bi7ty) that his server was crashing! A bit of debugging on a Saturday, and I realized that we had introduced a regression into a code path we didn't often use ourselves. I filed [MM-21619](https://mattermost.atlassian.net/browse/MM-22619) and got to work on the fix. Let's find it in the `master` branch.
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+	th := Setup(t)
+=======
+	th := SetupWithCustomConfig(t, func(config *model.Config) {
+		*config.PluginSettings.Enable = false
+	})
+>>>>>>> 9a51c73f6... MM-22619: check for nil plugins environment (#13901)
+	defer th.TearDown()
+```
+
+I have a mergetool configured with git to show this same information using a three-way merge inside Vim:
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
I have a mergetool configured with Git to show this same information using a three-way merge inside Vim:
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+
+The middle pane shows the common ancestor. On its left is what's in the current branch, and on the right is what's incoming. The bottom pane is the current file, showing the conflict markers. In general with merge conflicts, you have to understand the context of the code you are changing in order to make the decision on what version to keep, or whether to blend the changes somehow. In this case, I happen to know that a colleague of mine removed a redundant call to `InitBasic` in the master branch as part of a cleanup effort, but those clean up changes aren't in this release branch. So I need to combine the two changes together and put the call to `InitBasic` on top of the my original code. Then I'll get rid of the other sections and the conflict markers.
+
+With the conflicts resolved, I'm good to finish the cherry-pick!
+
+```sh
+git cherry-pick --continue
+```
+
+I've created a new commit that contains the changes from another branch, and would now be ready to push up this branch for review and CI testing before merging it into the release branch.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
I've created a new commit that contains the changes from another branch, and would now be ready to push up this branch for review and CI testing before merging it into the `release` branch.
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+
+The conflict resolution I just performed might have just as easily happened when doing a `git revert`. The process would be the same, and the hard part shouldn't be the tool you are using, but knowing the context of the code you are changing so you can make the right decisions.
+
+### git rebase
+
+My all time favourite git command is `git rebase`! Long before I learned git, I used a version control system called Perforce. It was undoubtedly powerful for its time, but I found it intensely frustrating to organize my code changes for review. Let me describe my typical workflow:
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
My all time favorite Git command is `git rebase`! Long before I learned Git, I used a version control system called Perforce. It was undoubtedly powerful for its time, but I found it intensely frustrating to organize my code changes for review. Let me describe my typical workflow:
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+```sh
+git checkout feature-branch
+vim ...
+git commit -m 'wip'
+vim ...
+git commit -m 'wip'
+...
+```
+
+This technique is known as "commit early and often." Every now and then, I'll also push my feature branch upstream (without opening a pull request) just to save my changes in case my computer dies. This is a perfectly fine way to develop, but I submit it makes for very poor pull requests. You can read my thoughts on [Submitting Great PRs](https://developers.mattermost.com/blog/submitting-great-prs/), but the short summary is that the best pull requests consist of a sequence of logically ordered commits that break up the changes into easy to follow chunks.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
This technique is known as "commit early and often". Every now and then, I'll also push my feature branch upstream (without opening a pull request) just to save my changes in case my computer dies. This is a perfectly fine way to develop, but it makes for very poor pull requests. You can read my thoughts on [Submitting Great PRs](https://developers.mattermost.com/blog/submitting-great-prs/), but the summary is that the best pull requests consist of a sequence of logically ordered commits that break up the changes into easy to follow chunks.
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+git checkout -b test-git-rebase
+cd model
+go test ./... -coverprofile=coverage.out && go tool cover -html=coverage.out -o coverage.html && open coverage.html
+```
+
+Let's take a look at the coverage of `model/config.go`:
+
+![image](/blog/2020-03-16-advanced-git-tbilisi-free-university/code-coverage-1.png)
+
+Ah, it appears we aren't explicitly adding a test to verify that the SiteURL is initialized differently when EnableDeveloper is true. Let's add that now.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
Ah, it appears we aren't explicitly adding a test to verify that the `SiteURL` is initialized differently when `EnableDeveloper` is `true`. Let's add that now.
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+
+Ok, save that and check again.
+
+```sh
+git add model/config.go
+git commit -m 'improve testing coe'
+go test ./... -coverprofile=coverage.out && go tool cover -html=coverage.out -o coverage.html && open coverage.html
+```
+
+Same coverage, but the tests are better structured. Let's see what else we can fix! Hmm, something's not right with EnableIncomingWebhooks:
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
Same coverage, but the tests are better structured. Let's see what else we can fix! Hmm, something's not right with `EnableIncomingWebhooks`:
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 14]
```diff
@@ -0,0 +1,437 @@
...
+    improve testing coe
+
+commit da0d04f00b9b11b10969a3467f88d109abc0ab9b
+Author: Jesse Hallam 
+Date:   Mon Mar 16 11:24:53 2020 -0300
+
+    add TestConfigEnableDeveloper
+```
+
+So the final diff is great, but the commits themselves are pretty disorganized. That's not how I want to present my work to my team. Imagine if these were dozens of commits over hundreds of changes. Imagine if I decided to rename a variable used all over the code half-way through my changes. It's pretty hard to review a set of disorganized changes. `git rebase` to the rescue! Let's run it interactively:
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
So the final diff is great, but the commits themselves are pretty disorganized. That's not how I want to present my work to my team. Imagine if these were dozens of commits over hundreds of changes. Imagine if I decided to rename a variable used all over the code halfway through my changes. It's pretty hard to review a set of disorganized changes so, `git rebase` to the rescue! Let's run it interactively:
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+It's thrown me into an editor, with each of my commits as a separate line. By default, they are all (cherry) "picked" to be kept, but I have a number of options to help me reorganize my code. So the first two commits and the last commit really don't need to be separated at all: they should just be a single commit. I'm going to move the last line up above the third commit, and then tell `git rebase` to squash the two commits into the first commit:
+
+```sh
+pick … (Jesse Hallam) add TestConfigEnableDeveloper
+squash … (Jesse Hallam) improve testing coe
+squash … (Jesse Hallam) wip
+pick  … (Jesse Hallam) remove dead code
+```
+
+When I save and exit this file, git will create a temporary ref at origin/master, and cherry-pick the changes in the order I requested. When I tell it to squash, commit amends to the previous commit, effectively squishing the three commits together. Save and exit, and you'll see I'm being prompted to write a new commit history for the squash commits. Finally, let's look at the diff:
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
When I save and exit this file, Git will create a temporary ref at `origin/master`, and cherry-pick the changes in the order I requested. When I tell it to squash, commit amends to the previous commit, effectively squishing the three commits together. Save and exit, and you'll see I'm being prompted to write a new commit history for the squash commits. Finally, let's look at the diff:
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 16]
```diff
@@ -0,0 +1,437 @@
...
+```sh
+git rebase -i origin/master
+```
+
+```sh
+reword … (Jesse Hallam) add TestConfigEnableDeveloper
+pick  … (Jesse Hallam) remove dead code
+```
+
+Neat! Of course, I didn't have to do this in two passes, but I wanted to show you that you can run it as often as necessary. Sometimes, when you reorder your commits, you'll end up with merge conflicts. Just as with `git cherry-pick` and `git revert`, `git rebase` will stop and ask you to resolve the conflicts. Then you can --continue and it will pick up with the next commit.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
Neat! Of course, I didn't have to do this in two passes, but I wanted to show you that you can run it as often as necessary. Sometimes, when you reorder your commits, you'll end up with merge conflicts. Just as with `git cherry-pick` and `git revert`, `git rebase` will stop and ask you to resolve the conflicts. Then you can `--continue` and it will pick up with the next commit.
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 17]
```diff
@@ -0,0 +1,437 @@
...
+
+Great! But... wait, I just deleted the commit! Oh no, my work is gone ... not exactly!
+
+### git reflog
+
+`git reflog` to the rescue! What is the reflog? https://git-scm.com/docs/git-reflog explains:
+
+> Reference logs, or 'reflogs', record when the tips of branches and other references were updated in the local repository.
+
+In layman terms, think of this way. Git remembers every commit you make, even if they aren't reachable by your named branches. Let's take a look at the reflog for HEAD:
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
In layman terms, think of this way: Git remembers every commit you make, even if they aren't reachable by your named branches. Let's take a look at the reflog for HEAD:
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 18]
```diff
@@ -0,0 +1,437 @@
...
+3489625af HEAD@{2020-03-16 17:53:19 -0300}: rebase -i (pick): wip
+f1ef494d4 HEAD@{2020-03-16 17:53:19 -0300}: rebase -i (pick): remove dead code
+b57fde469 HEAD@{2020-03-16 17:53:19 -0300}: rebase -i (pick): improve testing coe
+da0d04f00 HEAD@{2020-03-16 17:53:19 -0300}: rebase -i (pick): add TestConfigEnableDeveloper
+1c498996a (origin/master, origin/HEAD, master) HEAD@{2020-03-16 17:53:19 -0300}: rebase -i (start): checkout origin/master
+```
+
+And there's my missing commit reference! I can `git show f1ef494d4` it to confirm, and then bring it back into my branch with a simple `git cherry-pick f1ef494d4`.
+
+I presented `git reflog` last, since I think it's one of the harder commands to fundamentally understand: but this is probably the one command you should learn first, because you recover from almost any mistake you make with git using the reflog.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
I presented `git reflog` last, since I think it's one of the harder commands to fundamentally understand: but this is probably the one command you should learn first, because you recover from almost any mistake you make with Git using the reflog.
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 19]
```diff
@@ -0,0 +1,437 @@
...
+
+# Conclusion
+
+There's a ton more one could say about these commands, and even other advanced git commands. My hope is that you won't give up if you find git intimidating, especially when you know you can rely on the reflog to experiment without fear.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
There's a ton more one could say about these commands, and even other advanced Git commands. My hope is that you won't give up if you find Git intimidating, especially when you know you can rely on the reflog to experiment without fear.
```

**Comment by lieut-data** [comment]
Wow, thanks for such a thorough review, @justinegeffen! I never thought to capitalize Git before, but it makes so much sense -- learned something new :)

**Comment by justinegeffen** [comment]
Absolute pleasure. :) I had to Google to confirm git v. Git. It's a bit of an odd one, being that the command is lowercase. I worked off the casing used here: https://try.github.io/ which I take as de facto. :)

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 20]
```diff
@@ -0,0 +1,437 @@
...
+	defer th.TearDown()
+```
+
+I have a mergetool configured with Git to show this same information using a three-way merge inside Vim:
+
+```sh
+git mergetool
+```
+
+The middle pane shows the common ancestor. On its left is what's in the current branch, and on the right is what's incoming. The bottom pane is the current file, showing the conflict markers. In general with merge conflicts, you have to understand the context of the code you are changing in order to make the decision on what version to keep, or whether to blend the changes somehow. In this case, I happen to know that a colleague of mine removed a redundant call to `InitBasic` in the master branch as part of a cleanup effort, but those clean up changes aren't in this release branch. So I need to combine the two changes together and put the call to `InitBasic` on top of the my original code. Then I'll get rid of the other sections and the conflict markers.
```
Comment (agarciamontoro) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]: Nit: I think that a screenshot of your editor with what you're describing here would be terrific! (and it's always cool to look at the editors of other developers!)

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 21]
```diff
@@ -0,0 +1,437 @@
...
+...
+```
+
+This technique is known as "commit early and often". Every now and then, I'll also push my feature branch upstream (without opening a pull request) just to save my changes in case my computer dies. This is a perfectly fine way to develop, but it makes for very poor pull requests. You can read my thoughts on [Submitting Great PRs](https://developers.mattermost.com/blog/submitting-great-prs/), but the summary is that the best pull requests consist of a sequence of logically ordered commits that break up the changes into easy to follow chunks.
+
+Let me walk you through an example. We're going to see about improving some code coverage for the model package:
+
+```sh
+# Checkout the mattermost-server repository (if not already done)
+git checkout https://github.com/mattermost/mattermost-server.git
```
Comment (agarciamontoro) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]: Same comment as above.

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 22]
```diff
@@ -0,0 +1,437 @@
...
+It's thrown me into an editor, with each of my commits as a separate line. By default, they are all (cherry) "picked" to be kept, but I have a number of options to help me reorganize my code. So the first two commits and the last commit really don't need to be separated at all: they should just be a single commit. I'm going to move the last line up above the third commit, and then tell `git rebase` to squash the two commits into the first commit:
+
+```sh
+pick … (Jesse Hallam) add TestConfigEnableDeveloper
+squash … (Jesse Hallam) improve testing coe
+squash … (Jesse Hallam) wip
+pick  … (Jesse Hallam) remove dead code
+```
+
+When I save and exit this file, Git will create a temporary ref at `origin/master`, and cherry-pick the changes in the order I requested. When I tell it to squash, commit amends to the previous commit, effectively squishing the three commits together. Save and exit, and you'll see I'm being prompted to write a new commit history for the squash commits. Finally, let's look at the diff:
```
Comment (agarciamontoro) [inline site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]: This is an English question actually: is squash and squish the same thing?

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 23]
```diff
@@ -0,0 +1,437 @@
...
+
+Just over a month later, one of larger customers upgraded to v4.9 and found their system grinding to a halt. We hopped on an emergency call to glean some information, and everything pointed to my changes making things much worse instead of much better. You can read the [details of my second investigation](https://community.mattermost.com/core/pl/n8bfq7wq77rkff8mgrrstqeuye) into what went wrong, but in the end, we decided to [revert my changes](https://github.com/mattermost/mattermost-server/pull/8659) and ship a patch release.
+
+When I undid these changes, I wanted to bring the code back to exactly the way it was before my changes. We already knew how that code behaved for this customer, and the goal was to restore stability by using the old code instead. I could, of course, make the changes manually: finding the old code and copying it over top of the new code. But one of Git's built-in commands is an operation to do this automatically. Let's see it in action!
+
+```sh
+# Checkout the mattermost-server repository
+git checkout https://github.com/mattermost/mattermost-server.git
```
[the remark this answers is not in the document]
Comment (lieut-data) [reply site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]: Yes! Good eye :)
```suggestion
git clone https://github.com/mattermost/mattermost-server.git
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 24]
```diff
@@ -0,0 +1,437 @@
...
+
+At Mattermost, `master` is our development branch. Every two months we [release a new version](https://developers.mattermost.com/contribute/getting-started/branching/) of Mattermost that branches off `master`. Our release team begins to qualify this release, inevitably finding issues. We first fix the bugs in `master`, and then cherry-pick those changes back to the `release` branch. Of course, there are lots of different ways of using Git to handle releases, but this is the strategy we use at Mattermost.
+
+![image](/blog/2020-03-16-advanced-git-tbilisi-free-university/branching-overview.png)
+
+Let's look at the most recent such feature release: v5.20.
+
+```sh
+# Checkout the mattermost-server repository (if not already done)
+git checkout https://github.com/mattermost/mattermost-server.git
```
[the remark this answers is not in the document]
Comment (lieut-data) [reply site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]: Nope, I never actually had to /run/ this for myself, and just messed up everywhere :)
```suggestion
git clone https://github.com/mattermost/mattermost-server.git
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md (hunk 21 above)
Comment (lieut-data) [reply site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]:
```suggestion
git clone https://github.com/mattermost/mattermost-server.git
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md (hunk 22 above)
Comment (lieut-data) [reply site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]: In English, yes, but you're absolutely right that we should use a more approachable term.
```suggestion
When I save and exit this file, Git will create a temporary ref at `origin/master`, and cherry-pick the changes in the order I requested. When I tell it to squash, commit amends to the previous commit, effectively squashing the three commits together. Save and exit, and you'll see I'm being prompted to write a new commit history for the squash commits. Finally, let's look at the diff:
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md [hunk 25]
```diff
@@ -0,0 +1,437 @@
...
+* If you're working out of a shared branch, your peers will pull the revert like any other commit. This is usually the right way to "undo changes" in a shared branch, vs. using a command that rewrites history and requires a force push instead.
+
+Of course, if you've ever done this before, you know that I've kind of cheated today: my revert was clean, in that there were no merge conflicts. Sometimes, especially if some time has passed, other changes might have touched the same lines, and the revert will require resolving those changes. Hold that thought, though, since I'll work through an example of resolving merge conflicts with the next command. 
+Remember, reverts are just new commits!
+
+### git cherry-pick
+
+Next up: cherry picking! The general idea of cherry picking -- not specific to Git -- is to pick and choose just what you want. You can imagine yourself standing in front of an actual cherry tree, and saying, "I'll take that one, and that one, and that one" and ignoring the rest.
+
+In git, the idea is similar. Given one or more commits -- anywhere in the tree -- apply those same changes to my current ref. Why would you want to do that?
```
[the remark this answers is not in the document]
Comment (lieut-data) [reply site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]: Agreed!
```suggestion
In git, the idea is similar. Given one or more commits -- anywhere in the tree -- apply those same changes to my current branch. Why would you want to do that?
```

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md (hunk 20 above)
Comment (lieut-data) [reply site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]: Good call! Let me get that re-setup.

File: site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md (hunk 22 above)
Comment (agarciamontoro) [reply site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md]: Cool, thanks!
